llvm/lib/SYCLLowerIR/SYCLVirtualFunctionsAnalysis.cpp

```diff
@@ -0,0 +1,117 @@
+//===--- SYCLVirtualFunctionsAnalysis.cpp ---------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains implementation of SYCLVirtualFunctionsAnalysis pass that
+// is responsible for checking that virtual functions are used properly in SYCL
+// device code:
+// - if a kernel submitted without calls_indirectly property performs virtual
+//   function calls, a diagnostic should be emitted
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLVirtualFunctionsAnalysis.h"
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SetVector.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/IR/DiagnosticInfo.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+
+namespace {
+
+using CallGraphTy = DenseMap<const Function *, SmallPtrSet<Value *, 8>>;
+
+void emitDiagnostic(const SmallVector<const Function *> &Stack) {
+  diagnoseSYCLIllegalVirtualFunctionCall(Stack);
+}
+
+void checkKernelImpl(const Function *F, const CallGraphTy &CG,
+                     SmallVector<const Function *> &Stack) {
+  Stack.push_back(F);
+  CallGraphTy::const_iterator It = CG.find(F);
+  // It could be that the function itself is a leaf and doesn't call anything
+  if (It != CG.end()) {
+    const SmallPtrSet<Value *, 8> &Callees = It->getSecond();
+    for (const Value *V : Callees) {
+      auto *Callee = dyn_cast<Function>(V);
+      if (Callee)
+        checkKernelImpl(Callee, CG, Stack);
+      else
```

Comment:
just confirming my understanding, we actually emit the error when `Callee` is nullptr, which can only happen when it's a virtual vall?

Reply:
Right. Note that `Callee` is only `nullptr` because it is not a `Function`, but something else (just a `Value`), i.e. the call graph itself does not contain any `nullptr` nodes.